File: apps/ios/Sources/OlloCrypto/CryptoEngine.swift
import Foundation

/// Production builds must back this protocol with official libsignal.
/// Do not implement a homegrown Double Ratchet here.
public protocol CryptoEngine: Sendable {
    func generateIdentity() throws -> IdentityMaterial
    func processPrekeyBundle(_ remote: Data) throws -> String
    func encrypt(sessionId: String, plaintext: Data) throws -> Data
    func decrypt(sessionId: String, payload: Data) throws -> Data
    func safetyNumber(local: Data, remote: Data) -> String
}

public struct IdentityMaterial: Sendable {
    public var identityX25519: Data
    public var identityEd25519: Data
    public var signedPrekey: Data
    public var signature: Data
    public var oneTimePrekeys: [Data]

    public init(
        identityX25519: Data,
        identityEd25519: Data,
        signedPrekey: Data,
        signature: Data,
        oneTimePrekeys: [Data]
    ) {
        self.identityX25519 = identityX25519
        self.identityEd25519 = identityEd25519
        self.signedPrekey = signedPrekey
        self.signature = signature
        self.oneTimePrekeys = oneTimePrekeys
    }
}

/// Fails closed until a libsignal-backed engine is bound. Shipping this
/// type as the production engine is a release blocker.
public struct UnboundCryptoEngine: CryptoEngine {
    public init() {}

    public func generateIdentity() throws -> IdentityMaterial {
        throw EngineError.unbound
    }

    public func processPrekeyBundle(_ remote: Data) throws -> String {
        throw EngineError.unbound
    }

    public func encrypt(sessionId: String, plaintext: Data) throws -> Data {
        throw EngineError.unbound
    }

    public func decrypt(sessionId: String, payload: Data) throws -> Data {
        throw EngineError.unbound
    }

    public func safetyNumber(local: Data, remote: Data) -> String {
        SafetyNumber.of(identityA: local, identityB: remote).digits
    }

    public enum EngineError: Error {
        case unbound
    }
}

File: apps/android/core/crypto/src/main/java/app/ollo/crypto/CryptoEngine.kt
package app.ollo.crypto

/**
 * Client-facing crypto port. Production implementation MUST be official
 * libsignal (`org.signal:libsignal-client`). Do not implement a homegrown
 * Double Ratchet behind this interface.
 *
 * The TypeScript engine in `packages/crypto` is a review-required reference
 * for web/tests and is not used on Android release builds.
 */
interface CryptoEngine {
    fun generateIdentity(): IdentityMaterial
    fun deviceRegistrationJson(name: String, platform: String): String
    fun processPrekeyBundle(remote: ByteArray): SessionHandle
    fun encrypt(session: SessionHandle, plaintext: ByteArray): ByteArray
    fun decrypt(session: SessionHandle, payload: ByteArray): ByteArray
    fun safetyNumber(localIdentity: ByteArray, remoteIdentity: ByteArray): String
}

data class IdentityMaterial(
    val identityX25519: ByteArray,
    val identityEd25519: ByteArray,
    val signedPrekey: ByteArray,
    val signedPrekeySignature: ByteArray,
    val oneTimePrekeys: List<ByteArray>,
)

class SessionHandle(val id: String)

/**
 * Fails closed until a libsignal-backed engine is bound. Shipping this
 * class as the production engine is a release blocker.
 */
class UnboundCryptoEngine : CryptoEngine {
    override fun generateIdentity(): IdentityMaterial =
        throw IllegalStateException("libsignal engine is not bound")

    override fun deviceRegistrationJson(name: String, platform: String): String =
        throw IllegalStateException("libsignal engine is not bound")

    override fun processPrekeyBundle(remote: ByteArray): SessionHandle =
        throw IllegalStateException("libsignal engine is not bound")

    override fun encrypt(session: SessionHandle, plaintext: ByteArray): ByteArray =
        throw IllegalStateException("libsignal engine is not bound")

    override fun decrypt(session: SessionHandle, payload: ByteArray): ByteArray =
        throw IllegalStateException("libsignal engine is not bound")

    override fun safetyNumber(localIdentity: ByteArray, remoteIdentity: ByteArray): String =
        SafetyNumber.of(localIdentity, remoteIdentity).digits
}
